%YAML 1.2
# SPDX-FileCopyrightText: NVIDIA CORPORATION & AFFILIATES
# Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0
---
name: input_compositor
components:
- name: image_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: image_in
    min_size: 1
- name: cam_info_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: cam_info_in
    min_size: 1
- name: out
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: out
    min_size: 1
- name: compositor
  type: nvidia::isaac_ros::MessageCompositor
  parameters:
    image_rx: image_in
    camera_model_rx: cam_info_in
    tx: out
---
name: bbox
components:
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
      capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- name: allocator
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 0
    block_size: 32
    num_blocks: 40
- type: nvidia::isaac::tensor_ops::BBoxGenerator
  parameters:
    image_width: 0
    image_height: 0
    bbox_width: 0
    bbox_height: 0
    bbox_loc_x: 0
    bbox_loc_y: 0
    bbox_count: 1
    pool: allocator
    transmitter: data_transmitter
---
name: crop_and_resizer
components:
- name: image_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: image_receiver
    min_size: 1
- name: bbox_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: bbox_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- name: allocator
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 1566720
    num_blocks: 40
- name: crop_and_resize_operator
  type: nvidia::isaac::tensor_ops::CropAndResize
  parameters:
    output_width: 0
    output_height: 0
    interp_type: "linear"
    keep_aspect_ratio: true
    receiver: image_receiver
    receiver_bbox: bbox_receiver
    transmitter: data_transmitter
    pool: allocator
    input_adapter: resource/adapter
    output_adapter: resource/adapter
    output_name: "image"
---
name: splitter
components:
- name: data_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver
    min_size: 1
- name: image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: image_transmitter
    min_size: 1
- name: camera_model_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: camera_model_transmitter
    min_size: 1
- name: splitter
  type: nvidia::isaac_ros::MessageSplitter
  parameters:
    rx: data_receiver
    image_tx: image_transmitter
    camera_model_tx: camera_model_transmitter
---
name: image_sink
components:
- name: signal
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: signal
    min_size: 1
- name: sink
  type: nvidia::isaac_ros::MessageRelay
  parameters:
    source: signal
---
name: camera_info_sink
components:
- name: signal
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: signal
    min_size: 1
- name: sink
  type: nvidia::isaac_ros::MessageRelay
  parameters:
    source: signal
---
name: resource
components:
- name: adapter
  type: nvidia::isaac::tensor_ops::ImageAdapter
  parameters:
    message_type: "VideoBuffer"
---
components:
- name: edge1
  type: nvidia::gxf::Connection
  parameters:
    source: input_compositor/out
    target: crop_and_resizer/image_receiver
- name: edge2
  type: nvidia::gxf::Connection
  parameters:
    source: bbox/data_transmitter
    target: crop_and_resizer/bbox_receiver
- name: edge3
  type: nvidia::gxf::Connection
  parameters:
    source: crop_and_resizer/data_transmitter
    target: splitter/data_receiver
- name: edge4
  type: nvidia::gxf::Connection
  parameters:
    source: splitter/camera_model_transmitter
    target: camera_info_sink/signal
- name: edge5
  type: nvidia::gxf::Connection
  parameters:
    source: splitter/image_transmitter
    target: image_sink/signal
---
name: utils
components:
- name: clock
  type: nvidia::gxf::RealtimeClock
- type: nvidia::gxf::MultiThreadScheduler
  parameters:
    clock: clock
    stop_on_deadlock: false
    check_recession_period_ms: 1
    worker_thread_number: 2
    worker_thread_name_id: "nitros_crop_node"
- type: nvidia::gxf::JobStatistics
  parameters:
    clock: clock
